### PR TITLE
Local DP Modifier

### PR DESCRIPTION
**Pull request by mohammadnaseri**
### Description
Local DP, where each client is responsible to perform DP on its model parameters before sending them to the server. 

### Related issues/PRs

Merge after https://github.com/adap/flower/pull/2909
### Explanation
A new modifier class `LocalDpMod` is implemented where the user can create an instance of the class and pass DP hyperparameters. It performs clipping on the client's model updates, and then adds noise to the model parameters.

### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.dev/docs/writing-documentation.html)
- [ ] Update the changelog entry below
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### CODE REVIEW

**Review by panh99 — APPROVED** [review]

File: src/py/flwr/client/mod/localdp_mod.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        self, clipping_norm: float, sensitivity: float, epsilon: float, delta: float
+    ) -> None:
+        if clipping_norm <= 0:
+            raise ValueError("The clipping norm should be a positive value.")
+
+        if sensitivity < 0:
+            raise ValueError("The sensitivity should be a non-negative value.")
+
+        if epsilon < 0:
+            raise ValueError("The epsilon should be a non-negative value.")
```
Comment (danieljanes) [inline src/py/flwr/client/mod/localdp_mod.py:73]:
```suggestion
            raise ValueError("Epsilon should be a non-negative value.")
```

File: src/py/flwr/client/mod/localdp_mod.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+        if delta < 0:
+            raise ValueError("The delta should be a non-negative value.")
```
Comment (danieljanes) [inline src/py/flwr/client/mod/localdp_mod.py:76]:
```suggestion
            raise ValueError("Delta should be a non-negative value.")
```

File: src/py/flwr/client/mod/localdp_mod.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        """
+        if msg.metadata.message_type != MessageType.TRAIN:
+            return call_next(msg, ctxt)
+
+        fit_ins = compat.recordset_to_fitins(msg.content, keep_input=True)
+        server_to_client_params = parameters_to_ndarrays(fit_ins.parameters)
+
+        # Call inner app
+        out_msg = call_next(msg, ctxt)
+        fit_res = compat.recordset_to_fitres(out_msg.content, keep_input=True)
```
Comment (danieljanes) [inline src/py/flwr/client/mod/localdp_mod.py:110]: In between those lines, we need to check whether `out_msg.has_content()`:
```suggestion
        out_msg = call_next(msg, ctxt)
        fit_res = compat.recordset_to_fitres(out_msg.content, keep_input=True)
```

File: doc/source/how-to-use-differential-privacy.rst [hunk 4]
```diff
@@ -92,4 +92,32 @@ In addition to the server-side strategy wrapper, the :code:`ClientApp` needs to
   )
 
 
+Local Differential Privacy
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+To utilize local differential privacy (DP) and add noise to the client model parameters before transmitting them to the server in Flower, users can employ the `LocalDpMod`. They are required to provide the following hyperparameters: clipping norm value, sensitivity, epsilon, and delta.
```
Comment (danieljanes) [inline doc/source/how-to-use-differential-privacy.rst:97]:
```suggestion
To utilize local differential privacy (DP) and add noise to the client model parameters before transmitting them to the server in Flower, you can use the `LocalDpMod`. The following hyperparameters need to be set: clipping norm value, sensitivity, epsilon, and delta.
```

File: doc/source/how-to-use-differential-privacy.rst [hunk 5]
```diff
@@ -92,4 +92,32 @@ In addition to the server-side strategy wrapper, the :code:`ClientApp` needs to
...
+
+.. image:: ./_static/DP/localdp.png
+  :align: center
+  :width: 700
+  :alt: local DP mod
+
+Below is a sample code about how to use :code:`LocalDpMod`:
```
Comment (danieljanes) [inline doc/source/how-to-use-differential-privacy.rst:104]:
```suggestion
Below is a code example that shows how to use :code:`LocalDpMod`:
```

File: doc/source/how-to-use-differential-privacy.rst [hunk 6]
```diff
@@ -92,4 +92,32 @@ In addition to the server-side strategy wrapper, the :code:`ClientApp` needs to
...
+
+.. code-block:: python
+
+  # Client-side
```
Comment (danieljanes) [inline doc/source/how-to-use-differential-privacy.rst:108]:
```suggestion
```

File: doc/source/how-to-use-differential-privacy.rst [hunk 7]
```diff
@@ -92,4 +92,32 @@ In addition to the server-side strategy wrapper, the :code:`ClientApp` needs to
...
+  from flwr.client.mod.localdp_mod import LocalDpMod
+
+  # Create an instance of the mod with the required params
+  local_dp_obj = LocalDpMod(
+      cfg.clipping_norm, cfg.sensitivity, cfg.epsilon, cfg.delta
+  )
+  # Add local_dp_obj to the client-side mods
+  app = fl.client.ClientApp(client_fn=client_fn, mods=[local_dp_obj])
```
Comment (danieljanes) [inline doc/source/how-to-use-differential-privacy.rst:116]:
```suggestion
  # Add local_dp_obj to the client-side mods

  app = fl.client.ClientApp(
    client_fn=client_fn,
    mods=[local_dp_obj],
  )
```

File: doc/source/how-to-use-differential-privacy.rst [hunk 8]
```diff
@@ -92,4 +92,32 @@ In addition to the server-side strategy wrapper, the :code:`ClientApp` needs to
...
+
+
 Please note that the order of mods, especially those that modify parameters, is important when using multiple modifiers. Typically, differential privacy (DP) modifiers should be the last to operate on parameters.
+
+Local Training using Privacy Engines
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+For ensuring data instance-level privacy during local model training on the client side, consider leveraging privacy engines such as Opacus and TensorFlow Privacy. For examples of using Flower with these engines, please refer to the Flower examples directory.
```
Comment (danieljanes) [inline doc/source/how-to-use-differential-privacy.rst:123]: We can link to the respective code examples here, or to the examples documentation more broadly.

File: src/py/flwr/client/mod/localdp_mod.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+from flwr.common.message import Message
+
+
+class LocalDpMod:
+    """Modifier for local differential privacy.
+
+    This mod clips the client model updates and
+    add noise to the params before sending them to the server.
+
+    It operates on messages with type MessageType.TRAIN.
```
Comment (danieljanes) [inline src/py/flwr/client/mod/localdp_mod.py:36]:
```suggestion
    It operates on messages of type `MessageType.TRAIN`.
```

File: src/py/flwr/client/mod/localdp_mod.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    compute_clip_model_update,
+)
+from flwr.common.message import Message
+
+
+class LocalDpMod:
+    """Modifier for local differential privacy.
+
+    This mod clips the client model updates and
+    add noise to the params before sending them to the server.
```
Comment (danieljanes) [inline src/py/flwr/client/mod/localdp_mod.py:34]:
```suggestion
    This mod clips the client model updates and
    adds noise to the params before sending them to the server.
```

**Review by danieljanes — APPROVED** [review]

**Pull request merged**